File: mobile/src/blueGecko/java/com/siliconlabs/bledemo/activity/HealthThermometerActivity.java
package com.siliconlabs.bledemo.activity;

import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattService;
import android.content.Intent;
import android.os.Bundle;

import androidx.appcompat.widget.Toolbar;

import android.text.TextUtils;
import android.util.Pair;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.siliconlabs.bledemo.Base.Activities.BaseActivity;
import com.siliconlabs.bledemo.R;
import com.siliconlabs.bledemo.Bluetooth.BLE.BlueToothService;
import com.siliconlabs.bledemo.Bluetooth.BLE.GattCharacteristic;
import com.siliconlabs.bledemo.Bluetooth.BLE.GattService;
import com.siliconlabs.bledemo.Bluetooth.BLE.TimeoutGattCallback;
import com.siliconlabs.bledemo.fragment.SelectDeviceDialog;
import com.siliconlabs.bledemo.models.TemperatureReading;
import com.siliconlabs.bledemo.Utils.BLEUtils;

import java.util.Arrays;
import java.util.List;

import butterknife.ButterKnife;
import butterknife.InjectView;

public class HealthThermometerActivity extends BaseActivity {
    private BlueToothService service;
    private boolean serviceHasBeenSet;
    private BlueToothService.Binding bluetoothBinding;
    private HealthThermometerActivityFragment healthThermometerActivityFragment;
    @InjectView(R.id.toolbar)
    Toolbar toolbar;
    private TemperatureReading.HtmType htmType = TemperatureReading.HtmType.UNKNOWN;
    private final TimeoutGattCallback gattCallback = new TimeoutGattCallback() {

        @Override
        public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
            super.onConnectionStateChange(gatt, status, newState);
            if (newState == BluetoothGatt.STATE_DISCONNECTED) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        onDeviceDisconnect();
                    }
                });
            }
        }

        @Override
        public void onServicesDiscovered(BluetoothGatt gatt, int status) {
            super.onServicesDiscovered(gatt, status);
            boolean startNotificationForCharacteristicFromHere = true;
            List<BluetoothGattService> services = gatt.getServices();
            if (services != null) {
                for (BluetoothGattService s : services) {
                    if (s.getCharacteristics() != null) {
                        for (BluetoothGattCharacteristic ch : s.getCharacteristics()) {
                            if (GattCharacteristic.TemperatureType.uuid.equals(ch.getUuid())) {
                                startNotificationForCharacteristicFromHere = false;
                                gatt.readCharacteristic(ch);
                                break;
                            }

                        }
                    }
                }
            }

            if (startNotificationForCharacteristicFromHere) {
                BLEUtils.SetNotificationForCharacteristic(gatt, GattService.HealthThermometer,
                        GattCharacteristic.Temperature,
                        BLEUtils.Notifications.INDICATE);
            }

        }

        @Override
        public void onCharacteristicChanged(final BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
            super.onCharacteristicChanged(gatt, characteristic);
            if (GattCharacteristic.fromUuid(characteristic.getUuid()) == GattCharacteristic.Temperature) {
                final TemperatureReading reading = TemperatureReading.fromCharacteristic(characteristic);
                reading.setHtmType(htmType);
                String deviceName = gatt.getDevice().getName();
                if (TextUtils.isEmpty(deviceName)) {
                    deviceName = gatt.getDevice().getAddress();
                }
                final String finalDeviceName = deviceName;
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        healthThermometerActivityFragment.setCurrentReading(reading);
                        healthThermometerActivityFragment.setDeviceName(finalDeviceName);
                    }
                });
            }
        }

        @Override
        public void onCharacteristicRead(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status) {
            super.onCharacteristicRead(gatt, characteristic, status);
            if (GattCharacteristic.fromUuid(characteristic.getUuid()) == GattCharacteristic.TemperatureType) {
                htmType = TemperatureReading.HtmType.values()[characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 0)];

                BLEUtils.SetNotificationForCharacteristic(gatt, GattService.HealthThermometer,
                        GattCharacteristic.Temperature,
                        BLEUtils.Notifications.INDICATE);
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_thermometer);
        ButterKnife.inject(this);
        setSupportActionBar(toolbar);

        findViewById(R.id.iv_go_back).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackPressed();
            }
        });

        healthThermometerActivityFragment = (HealthThermometerActivityFragment) getSupportFragmentManager().findFragmentById(R.id.fragment);
        bluetoothBinding = new BlueToothService.Binding(this) {
            @Override
            protected void onBound(final BlueToothService service) {
                HealthThermometerActivity.this.service = service;
                serviceHasBeenSet = true;
                if (!service.isGattConnected()) {
                    Toast.makeText(HealthThermometerActivity.this, R.string.toast_htm_gatt_conn_failed, Toast.LENGTH_LONG).show();

                    if (service != null) {
                        service.clearGatt();
                    }
                    bluetoothBinding.unbind();

                    finish();
                } else {
                    service.registerGattCallback(true, gattCallback);
                    service.discoverGattServices();
                }
            }
        };
        BlueToothService.bind(bluetoothBinding);
    }

    @Override
    protected void onResume() {
        super.onResume();
        //get out if the service has stopped, or if the gatt connection is dead
        if ((serviceHasBeenSet && service == null) || (service != null && !service.isGattConnected())) {
            onDeviceDisconnect();
        }
    }

    private void onDeviceDisconnect() {
        if (!isFinishing()) {
            Toast.makeText(HealthThermometerActivity.this, R.string.device_has_disconnected, Toast.LENGTH_SHORT).show();
            finish();
        }
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        //When you switch thermometers, this activity should get a new intent - at this point, we hide the dialog
        SelectDeviceDialog dialog = (SelectDeviceDialog) getSupportFragmentManager().findFragmentByTag("select_device_tag");
        if (dialog != null && dialog.isVisible()) {
            dialog.dismiss();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (service != null) {
            service.clearGatt();
        }
        bluetoothBinding.unbind();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (item.getItemId() == android.R.id.home) {
            if (service != null && service.getConnectedGatt() != null)
                BLEUtils.SetNotificationForCharacteristic(service.getConnectedGatt(), GattService.HealthThermometer, GattCharacteristic.Temperature,
                        BLEUtils.Notifications.DISABLED);

            finish();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public void onChangeButtonClick() {
        SelectDeviceDialog dialog = SelectDeviceDialog.newDialog(R.string.title_Health_Thermometer, R.string.main_menu_description_thermometer,
                Arrays.asList(new Pair<>(R.string.htp_title, R.string.htp_id)), BlueToothService.GattConnectType.THERMOMETER);
        dialog.show(getSupportFragmentManager(), "select_device_tag");
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/health_thermometer/activities/HealthThermometerActivity.kt
package com.siliconlabs.bledemo.health_thermometer.activities

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import com.siliconlabs.bledemo.base.BaseActivity
import com.siliconlabs.bledemo.base.SelectDeviceDialog
import com.siliconlabs.bledemo.bluetooth.services.BluetoothService
import com.siliconlabs.bledemo.bluetooth.ble.GattCharacteristic
import com.siliconlabs.bledemo.bluetooth.ble.GattService
import com.siliconlabs.bledemo.bluetooth.ble.TimeoutGattCallback
import com.siliconlabs.bledemo.health_thermometer.fragments.HealthThermometerActivityFragment
import com.siliconlabs.bledemo.health_thermometer.models.TemperatureReading
import com.siliconlabs.bledemo.health_thermometer.models.TemperatureReading.HtmType
import com.siliconlabs.bledemo.R
import com.siliconlabs.bledemo.utils.BLEUtils
import com.siliconlabs.bledemo.utils.BLEUtils.setNotificationForCharacteristic
import kotlinx.android.synthetic.main.actionbar.*

class HealthThermometerActivity : BaseActivity() {
    private lateinit var healthThermometerActivityFragment: HealthThermometerActivityFragment
    private lateinit var bluetoothBinding: BluetoothService.Binding
    private var service: BluetoothService? = null
    private var htmType = HtmType.UNKNOWN
    private var serviceHasBeenSet = false

    private val bluetoothAdapterStateChangeListener: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            if (action == BluetoothAdapter.ACTION_STATE_CHANGED) {
                val state = intent.getIntExtra(BluetoothAdapter.EXTRA_STATE, BluetoothAdapter.ERROR)
                if (state == BluetoothAdapter.STATE_OFF) finish()
            }
        }
    }

    private val gattCallback: TimeoutGattCallback = object : TimeoutGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            super.onConnectionStateChange(gatt, status, newState)
            if (newState == BluetoothGatt.STATE_DISCONNECTED) {
                runOnUiThread { onDeviceDisconnect() }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            super.onServicesDiscovered(gatt, status)
            var startNotificationForCharacteristicFromHere = true
            val services = gatt.services
            if (services != null) {
                for (s in services) {
                    if (s.characteristics != null) {
                        for (ch in s.characteristics) {
                            if (GattCharacteristic.TemperatureType.uuid == ch.uuid) {
                                startNotificationForCharacteristicFromHere = false
                                gatt.readCharacteristic(ch)
                                break
                            }
                        }
                    }
                }
            }
            if (startNotificationForCharacteristicFromHere) {
                setNotificationForCharacteristic(gatt, GattService.HealthThermometer,
                        GattCharacteristic.Temperature,
                        BLEUtils.Notifications.INDICATE)
            }
        }

        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
            super.onCharacteristicChanged(gatt, characteristic)
            if (GattCharacteristic.fromUuid(characteristic.uuid) == GattCharacteristic.Temperature) {
                val reading = TemperatureReading.fromCharacteristic(characteristic)
                reading.htmType = htmType
                var deviceName = gatt.device.name
                if (TextUtils.isEmpty(deviceName)) {
                    deviceName = gatt.device.address
                }
                val finalDeviceName = deviceName
                runOnUiThread {
                    healthThermometerActivityFragment.setCurrentReading(reading)
                    healthThermometerActivityFragment.setDeviceName(finalDeviceName)
                }
            }
        }

        override fun onCharacteristicRead(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
            super.onCharacteristicRead(gatt, characteristic, status)
            if (GattCharacteristic.fromUuid(characteristic.uuid) == GattCharacteristic.TemperatureType) {
                htmType = HtmType.values()[characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 0)]
                setNotificationForCharacteristic(gatt, GattService.HealthThermometer,
                        GattCharacteristic.Temperature,
                        BLEUtils.Notifications.INDICATE)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_thermometer)
        prepareToolbar()

        val filter = IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED)
        registerReceiver(bluetoothAdapterStateChangeListener, filter)

        healthThermometerActivityFragment = supportFragmentManager.findFragmentById(R.id.fragment) as HealthThermometerActivityFragment

        bluetoothBinding = object : BluetoothService.Binding(this) {
            override fun onBound(service: BluetoothService?) {
                this@HealthThermometerActivity.service = service
                serviceHasBeenSet = true
                if (!service?.isGattConnected()!!) {
                    showMessage(R.string.toast_htm_gatt_conn_failed)
                    service.clearConnectedGatt()
                    bluetoothBinding.unbind()
                    finish()
                } else {
                    service.registerGattCallback(true, gattCallback)
                    service.discoverGattServices()
                }
            }
        }

        bluetoothBinding?.bind()
    }

    private fun prepareToolbar() {
        setSupportActionBar(toolbar)
        findViewById<View>(R.id.iv_go_back).setOnClickListener { onBackPressed() }
    }

    override fun onResume() {
        super.onResume()
        if (serviceHasBeenSet && service == null || service != null && !service?.isGattConnected()!!) {
            onDeviceDisconnect()
        }
    }

    private fun onDeviceDisconnect() {
        if (!isFinishing) {
            showMessage(R.string.device_has_disconnected)
            finish()
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        //When you switch thermometers, this activity should get a new intent - at this point, we hide the dialog
        val dialog = supportFragmentManager.findFragmentByTag("select_device_tag") as SelectDeviceDialog?
        if (dialog != null && dialog.isVisible) {
            dialog.dismiss()
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        unregisterReceiver(bluetoothAdapterStateChangeListener)
        service?.clearConnectedGatt()
        bluetoothBinding.unbind()
    }
}